File: rust/src/models/add_on_event_data.rs
// this file is @generated
use serde::{Deserialize, Serialize};

use super::{
    add_on_id::AddOnId, price_id::PriceId, product_fee_type_enum::ProductFeeTypeEnum,
    product_id::ProductId,
};

#[derive(Clone, Debug, Default, PartialEq, Deserialize, Serialize)]
pub struct AddOnEventData {
    pub add_on_id: AddOnId,

    pub created_at: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub fee_type: Option<ProductFeeTypeEnum>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_instances_per_subscription: Option<i32>,

    pub name: String,

    pub price_id: PriceId,

    pub product_id: ProductId,

    pub self_serviceable: bool,
}

impl AddOnEventData {
    pub fn new(
        add_on_id: AddOnId,
        created_at: String,
        name: String,
        price_id: PriceId,
        product_id: ProductId,
        self_serviceable: bool,
    ) -> Self {
        Self {
            add_on_id,
            created_at,
            description: None,
            fee_type: None,
            max_instances_per_subscription: None,
            name,
            price_id,
            product_id,
            self_serviceable,
        }
    }
}

File: typescript/src/models/addOnEventData.ts
// this file is @generated
import { parseDateTime } from "../datetime";
import { type AddOnId, AddOnIdSerializer } from "./addOnId";
import { type PriceId, PriceIdSerializer } from "./priceId";
import {
  type ProductFeeTypeEnum,
  ProductFeeTypeEnumSerializer,
} from "./productFeeTypeEnum";
import { type ProductId, ProductIdSerializer } from "./productId";

export interface AddOnEventData {
  addOnId: AddOnId;

  createdAt: Date;

  description?: string | null;

  feeType?: ProductFeeTypeEnum | null;

  maxInstancesPerSubscription?: number | null;

  name: string;

  priceId: PriceId;

  productId: ProductId;

  selfServiceable: boolean;
}

export const AddOnEventDataSerializer = {
  _fromJsonObject(object: any): AddOnEventData {
    return {
      addOnId: AddOnIdSerializer._fromJsonObject(object["add_on_id"]),
      createdAt: parseDateTime(object["created_at"]),
      description: object["description"],
      feeType:
        object["fee_type"] != null
          ? ProductFeeTypeEnumSerializer._fromJsonObject(object["fee_type"])
          : undefined,
      maxInstancesPerSubscription: object["max_instances_per_subscription"],
      name: object["name"],
      priceId: PriceIdSerializer._fromJsonObject(object["price_id"]),
      productId: ProductIdSerializer._fromJsonObject(object["product_id"]),
      selfServiceable: object["self_serviceable"],
    };
  },

  _toJsonObject(self: AddOnEventData): any {
    return {
      add_on_id: AddOnIdSerializer._toJsonObject(self.addOnId),
      created_at: self.createdAt,
      description: self.description,
      fee_type:
        self.feeType != null
          ? ProductFeeTypeEnumSerializer._toJsonObject(self.feeType)
          : undefined,
      max_instances_per_subscription: self.maxInstancesPerSubscription,
      name: self.name,
      price_id: PriceIdSerializer._toJsonObject(self.priceId),
      product_id: ProductIdSerializer._toJsonObject(self.productId),
      self_serviceable: self.selfServiceable,
    };
  },
};
